pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,459 @@
+	// https://github.com/libp2p/go-libp2p-pubsub/blob/cbb7bfc1f182e0b765d2856f6a0ea73e34d93602/tracer.go#L280
+	var s network.Stream
+	var err error
+	for i := 0; i < maxRetries; i++ {
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, err = t.host.NewStream(sctx, p, ProtocolID)
+		cancel()
+
+		if err == nil {
+			return s, nil
+		}
+
+		if i >= noLoggingErrorAttempts {
+			t.logger.Log(
+				logging.LevelError, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, err))
+		} else {
+			t.logger.Log(
+				logging.LevelInfo, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, err))
+		}
+
+		delay := time.NewTimer(maxRetryTimeout)
+		select {
+		case <-delay.C:
+			continue
+		case <-ctx.Done():
+			if !delay.Stop() {
+				<-delay.C
+			}
+			return nil, fmt.Errorf("%s opening stream to %s: context closed", t.ownID, dest)
+		}
+	}
+	return nil, fmt.Errorf("%s failed to open stream to %s: %w", t.ownID, dest, err)
+}
+
+func (t *Transport) sendPayload(dest types.NodeID, payload []byte) error {
+	t.connsLock.RLock()
+	defer t.connsLock.RUnlock()
+
+	// There are two cases when we get an error:
+	// 1. We don't have the node ID in the nodes table. E.g. we didn't call Connect().
+	// 2. We created an entry for the node but have not opened a connection.
+	// But if we added the node via Connect, then it should open the connection.
+	// It doesn't make sense to open a new one here.
+	s, err := t.getNodeStreamWithoutLock(dest)
+	if err != nil {
+		return err
+	}
+
+	if _, err := s.Write(payload); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *Transport) encode(msg *messagepb.Message) ([]byte, error) {
+	p, err := proto.Marshal(msg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	tm := TransportMessage{t.ownID.Pb(), p}
+	buf := new(bytes.Buffer)
+	if err = tm.MarshalCBOR(buf); err != nil {
+		return nil, fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+	return buf.Bytes(), nil
+}
+
+func (t *Transport) readAndDecode(s io.Reader) (*messagepb.Message, types.NodeID, error) {
+	var tm TransportMessage
+	err := tm.UnmarshalCBOR(s)
+	if err != nil {
+		return nil, "", err
+	}
+
+	var msg messagepb.Message
+	if err := proto.Unmarshal(tm.Payload, &msg); err != nil {
+		return nil, "", err
+	}
+	return &msg, types.NodeID(tm.Sender), nil
+}
+
+func (t *Transport) mirHandler(s network.Stream) {
+	peerID := s.Conn().RemotePeer().String()
+
+	t.logger.Log(logging.LevelDebug, "mir handler started", "src", t.ownID, "dst", peerID)
+	defer t.logger.Log(logging.LevelDebug, "mir handler stopped", "src", t.ownID, "dst", peerID)
+
+	defer func() {
+		t.logger.Log(logging.LevelDebug, "mir handler is closing stream", "src", t.ownID, "dst", peerID)
+		err := s.Close()
+		if err != nil {
+			t.logger.Log(logging.LevelError, "closing stream", "src", t.ownID, "dst", peerID, "err", err)
+		}
+	}() // nolint
+
+	for {
+		msg, sender, err := t.readAndDecode(s)
```

Comment:
What "preemptable call" is? The call using gosched?

Reply:
Added stopChan. Did you mean something like that? if you did not let's discuss

Reply:
Yes, by "preemptable" I meant that we must be able to interrupt the waiting for the data. The `stopChan` is a way of achieving that (since when `stopChan` is closed, the stream is closed, which makes the unmarshalling return, which in turn makes `readAndDecode` return).